### PR TITLE
Upgrade FSDP checkpointing to be compatible with torch 2.0

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?
1. Updates how we materialize optimizer state_dicts and load them with FSDP to be compatible with torch 2.0
2. Removes some unnecessary unit tests

# What issue(s) does this change relate to?
fix CO-2041

### CODE REVIEW

File: composer/core/state.py [hunk 1]
```diff
@@ -119,17 +119,9 @@ def fsdp_get_optim_state_dict(model: torch.nn.Module,
...
-        return FSDP.full_optim_state_dict(model=model, optim=optim)
-    elif state_dict_type == 'sharded':
-        # Converts local state dict to sharded.
-        return FSDP.sharded_optim_state_dict(model=model, optim=optim)
-    elif state_dict_type == 'local':
-        # State dict is already local, so just return state dict.
-        return optim.state_dict()
-    else:
-        raise NotImplementedError(f'No valid FSDP state_dict_type for {state_dict_type}')
+    with fsdp_state_dict_type_context(module=model, state_dict_type=state_dict_type):
```
Comment (mvpatel2000) [inline composer/core/state.py:122]: does this work w torch 1.13

**Review by mvpatel2000 — COMMENTED** [review]
why are the tests deleted

File: composer/core/state.py (hunk 1 above)
Comment (dakinggg) [reply composer/core/state.py:122]: +1

File: composer/core/state.py [hunk 2]
```diff
@@ -950,29 +942,15 @@ def load_optim_state(self, state_dict: Dict[str, Any]):
...
-                    # Optimizer and optimizer state dict are already sharded, but not
-                    # flattened, so we flatten the state dict then load it.
-                    flattened_optim_state_dict = FSDP.flatten_sharded_optim_state_dict(
-                        sharded_optim_state_dict=optim_state_dict, model=self.model, optim=optimizer)
-                    optimizer.load_state_dict(flattened_optim_state_dict)
-                elif self.fsdp_state_dict_type == 'local':
-                    # Optimizer and optimizer state dict are already sharded and flattened,
-                    # so just load the state_dict.
+                with fsdp_state_dict_type_context(module=self.model, state_dict_type=self.fsdp_state_dict_type):
+                    optim_state_dict = FSDP.optim_state_dict_to_load(optim_state_dict=optim_state_dict,
```
Comment (dakinggg) [inline composer/core/state.py:950]: same q here, does it work with torch 1.13?

File: tests/trainer/test_sharded_checkpoint.py [hunk 3]
```diff
@@ -105,99 +116,6 @@ def _compare_model_params_between_state_dicts(state_dict1, state_dict2):
                            state_dict2_model_tensor), f'Weight named {param_name} not the same between state_dicts'
 
 
-@pytest.mark.gpu
```
Comment (dakinggg) [inline tests/trainer/test_sharded_checkpoint.py:108]: could you please justify/explain deleting these tests?

File: composer/core/state.py (hunk 2 above)
Comment (eracah) [reply composer/core/state.py]: no, but fixed now

File: composer/core/state.py (hunk 1 above)
Comment (eracah) [reply composer/core/state.py]: no, but fixed now

File: tests/trainer/test_sharded_checkpoint.py (hunk 3 above)
Comment (eracah) [reply tests/trainer/test_sharded_checkpoint.py:108]: we talked offline. they aren't truly necessary to test functionality, but could be good for picking up on some weird torch warnings. will try to add back in

File: composer/core/state.py [hunk 4]
```diff
@@ -119,6 +119,20 @@ def fsdp_get_optim_state_dict(model: torch.nn.Module,
     if version.parse(torch.__version__) < version.parse('1.13.0'):
         raise RuntimeError('To use FSDP with Composer, you must use torch>=1.13.0.')
     from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+    if version.parse(torch.__version__) < version.parse('2.0.0'):
```
Comment (dakinggg) [inline composer/core/state.py:122]: use helper `using_torch_2_0()`

File: composer/core/state.py [hunk 5]
```diff
@@ -950,29 +988,22 @@ def load_optim_state(self, state_dict: Dict[str, Any]):
...
-                    optimizer.load_state_dict(optim_state_dict)
-                else:  # fsdp_state_dict_type == 'full'
-                    # FSDP enabled, but fsdp_state_dict is set to 'full', so the state dict
-                    # is a full state dict and we must shard and flatten it first before loading it.
-                    sharded_optim_state_dict = get_fsdp_sharded_optim_state_dict(full_optim_state_dict=optim_state_dict,
-                                                                                 model=self.model)
-                    log.debug(f'optimizer.load_state_dict call with fsdp_state_dict_type=full')
-                    optimizer.load_state_dict(sharded_optim_state_dict)
-            # No FSDP, so just load the optim state dict.
+                if version.parse(torch.__version__) < version.parse('2.0.0'):
```
Comment (dakinggg) [inline composer/core/state.py:995]: same

File: tests/trainer/test_sharded_checkpoint.py [hunk 6]
```diff
@@ -14,8 +14,18 @@
 from composer.algorithms import EMA
 from composer.trainer import Trainer
 from composer.utils import dist
-from tests.common import RandomClassificationDataset, SimpleModel
+from tests.common import RandomClassificationDataset
 from tests.common.markers import world_size
+from composer.models import ComposerClassifier
+
+class SimpleMLP(ComposerClassifier):
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply tests/trainer/test_sharded_checkpoint.py:22]: jk, keep this here, please add a comment saying why its in this file

File: tests/trainer/test_sharded_checkpoint.py [hunk 7]
```diff
@@ -245,6 +164,47 @@ def test_fsdp_full_state_dict_load(world_size, tmp_path: pathlib.Path, autoresum
...
+    if version.parse(torch.__version__) >= version.parse('2.0.0') and state_dict_type == 'local':
+        pytest.xfail("Loading a torch 1.13 checkpoint with torch 2.0 for state_dict_type local is not backwards compatible")
+    
+
+    rank = 0 if state_dict_type == 'full' else '{rank}'
+    load_path = f's3://mosaicml-internal-checkpoints-test/ckpt-compatibility-test/{sharding_strategy.lower()}_{state_dict_type}_{precision}/rank{rank}.pt'
+
+    trainer2 = get_trainer(
+        fsdp_state_dict_type=state_dict_type,
+        num_features=32,
```
Comment (dakinggg) [inline tests/trainer/test_sharded_checkpoint.py:185]: comment that these matter

**Review by dakinggg — APPROVED** [review]

**Pull request merged**